fe/fe-core/src/main/java/org/apache/doris/statistics/util/InternalSqlTemplate.java

```diff
@@ -0,0 +1,282 @@
+    public static final String PARTITION_NDV_VALUE_SQL = "SELECT NDV(${column}) AS ndv FROM"
+            + " ${table} PARTITION (${partition});";
+
+    public static final String MIN_MAX_NDV_VALUE_SQL = "SELECT MIN(${column}) AS min_value, MAX(${column})"
+            + " AS max_value, NDV(${column}) AS ndv FROM ${table};";
+    public static final String PARTITION_MIN_MAX_NDV_VALUE_SQL = "SELECT MIN(${column}) AS min_value,"
+            + " MAX(${column}) AS max_value, NDV(${column}) AS ndv FROM ${table} PARTITION (${partition});";
+
+    public static final String ROW_COUNT_SQL = "SELECT COUNT(1) AS row_count FROM ${table};";
+    public static final String PARTITION_ROW_COUNT_SQL = "SELECT COUNT(1) AS row_count FROM ${table} PARTITION"
+            + " (${partition});";
+
+    public static final String MAX_SIZE_SQL = "SELECT MAX(LENGTH(${column})) AS max_size FROM ${table};";
+    public static final String PARTITION_MAX_SIZE_SQL = "SELECT MAX(LENGTH(${column})) AS max_size FROM"
+            + " ${table} PARTITION (${partition});";
+
+    public static final String AVG_SIZE_SQL = "SELECT AVG(LENGTH(${column})) AS avg_size FROM ${table};";
+    public static final String PARTITION_AVG_SIZE_SQL = "SELECT AVG(LENGTH(${column})) AS avg_size"
+            + " FROM ${table} PARTITION (${partition});";
+
+    public static final String MAX_AVG_SIZE_SQL = "SELECT MAX(LENGTH(${column})) AS max_size,"
+            + " AVG(LENGTH(${column})) AS avg_size FROM ${table};";
+    public static final String PARTITION_MAX_AVG_SIZE_SQL = "SELECT MAX(LENGTH(${column}))"
+            + " AS max_size, AVG(LENGTH(${column})) AS avg_size FROM ${table} PARTITION (${partition});";
+
+    public static final String NUM_NULLS_SQL = "SELECT COUNT(1) AS num_nulls FROM ${table}"
+            + " WHERE ${column} IS NULL;";
+    public static final String PARTITION_NUM_NULLS_SQL = "SELECT COUNT(1) AS num_nulls FROM"
+            + " ${table} PARTITION (${partition}) WHERE ${column} IS NULL;";
+    /** ---------------------------- for statistics end ---------------------------- */
+
+    private static final Logger LOG = LogManager.getLogger(InternalSqlTemplate.class);
+
+    private static final Pattern PATTERN = Pattern.compile("\\$\\{\\w+\\}");
+
+    /**
+     * Concatenate SQL statements based on templates and parameters. e.g.:
+     * template and parameters:
+     *  'SELECT ${col} FROM ${table} WHERE id = ${id};',
+     *   parameters: {col=colName, table=tableName, id=1}
+     *   result sql: 'SELECT colName FROM tableName WHERE id = 1;
+     * <p>
+     *
+     * @param template sql template
+     * @param params   k,v parameter, if without parameter, params should be null
+     * @return SQL statement with parameters concatenated
+     */
+    public static String processTemplate(String template, Map<String, String> params) {
+        Matcher matcher = PATTERN.matcher(template);
+        StringBuffer sb = new StringBuffer();
```

Comment:
It's a local variable which means there is no race condition, so we don't need `StringBuffer` Here, `StringBuilder` is  much enough

Reply:
It seems that using StringBuilder makes the following appendReplacement() and appendTail() throw exceptions (prompting me to use jdk9++, Usage of API documented as @since 9++), so I used StringBuffer here.

Reply:
Ok